Reducing common structs types or https://github.com/kube-rs/gateway-api-rs/issues/38

dawid-nowak opened the pull request:
This approach aims to simplify the types and eliminate redundant ones. The approach works as follows:
1. We scan generated files and find similar items (structs or enums). Two structs are considered to be similar and processed further if their fields are exactly the same. The same method applies to the enums.
2. Once all similar structs and enums have been found, then we try get the name that will applied to similar structs from a file.
3. Once the new type names are settled we change the generated code and add a new file with the reduced types "common_types.rs"

We follow this process in phases:
1. In Phase 1, we only look at the structs with fields which are built from simple types (String, u32, u64)
2. In Phase 2, we also look at the structs with fields that have been changed in Phase 1
3. and repeat

Problems:
1. The documentation is stripped since the same type could be used in many different locations

dawid-nowak commented:
Thanks for the kind words. Let me answer your questions:

***UX***
I think the "processed" name is not great but that could be easily changed. The Kubvernor Gateway implementation was initially based on "standard" APIs but I wanted to see whether "processed" types would improve the code. I have migrated the existing codebase to "processed" APIs and I think [the results](https://github.com/kubvernor/kubvernor/tree/new_types_poc) are optimistic (at least from my perspective).

Initially, I created a lot of types manually to handle duplication between the HTTP and GRPC Routes. With "processed" types I was able to automate that part somewhat. As a result, I was able to (re)move and refactor some duplicated code. 
The one thing I found is that I would never mix the types. I would either rely on the types provided by "standard" APIs or on types provided by "processed" APIs. 

It is very doable to emit "common" types and keep them side by side with "standard" types. In this approach I think we would need to provide a whole bunch of converters from "standard" types into "common" types. This would be very much akin to my original approach to implementing Kubvernor. 

I think from the developer perspective such an approach would be confusing because the developers would need to keep selecting which type to use at a given moment in time and keep on converting from one type to another. 

***Maintainability & Kopium Interactions***
This is an interesting one and it prompted me to learn more about the Kubernetes APIs.
Ultimately, I think the problem is that the yaml files describing CRDs are not using OpenAPI reference mechanisms to avoid duplication of types. 

When one investigates the Go implementation, it is clear that there is a lot of code reuse there. GRPCRoute is dependent on types from HTTPRoute [for example](https://github.com/kubernetes-sigs/gateway-api/blob/392be7c235c4cd516a9765abe7dd30e9dca6c4ee/apis/v1/grpcroute_types.go#L540) and there is [shared_types](https://github.com/kubernetes-sigs/gateway-api/blob/392be7c235c4cd516a9765abe7dd30e9dca6c4ee/apis/v1/shared_types.go) module with type definitions shared among all CRDs. 

This reuse disappears when yaml versions of the CRDs are generated.  All information on how the types are reused is erased which makes it super hard for Kopium.

I think that if yaml/OpenAPI definitions relied on OpenAPI references to reuse shared definitions then it would be much easier for Kopium to generate Rust code very similar to the one in "processed" APIs. Indeed, this PR probably wouldn't even be necessary. 

I don't know enough the process to speculate whether it would be possible to change how yaml versions are generated.

shaneutt commented:
### Maintenance & Kopium

Using refs in the specification would be more ideal. We discussed this a bit in #38 and I personally think it's an inevitability in the long run if we're going to have strong support for non-Golang implementations. We can still move forward with something that works for us for now, however. The only conflict I would worry about it whether we try to mint a `v1` of this library knowing that this might change significantly down the road. For the purposes of this PR however, I don't think we sweat that, but what GA is going to look like for this library is always in the back of my mind.

### UX

I'm not sure I'm clear yet: can we adapt this work to emit less, and keep with the original structure?

dawid-nowak commented:
**Maintenance & Kopium**
Looks like this has been around for a while?
https://github.com/kubernetes/kubernetes/issues/62872

**UX**
Yeah, in terms of the UX I think I am not clear what the ask is either 😃 
I think it would be good if you said exactly what is the desired outcome from your perspective in the context of this PR. Then we can discuss whether it is possible, doable or how it fits with my project. I hope this sounds fair.

dawid-nowak commented:
I think exposing common types as `gateway_api::apis::standard::common::*` shouldn't be a problem. 

I am not clear on the first part. 

Using HTTPRoute as an example.

Even if we keep the preferred layout , `gateway_api::apis::standard::httproutes::HTTPRoute`  will be changed to use the types from `gateway_api::apis::standard::common::*`. This would be ideal but it completely breaks the existing clients.

Otherwise, we either need to create a new module so we can keep the name of the type as is or create a new type in `gateway_api::apis::standard::httproutes` 

Creating a new module is much simpler to implement. Interleaving types will be problematic due to the recursive nature of the algorithm. Also the types will have to have different names so they can co-exist.

shaneutt commented:
Your example was helpful. As far as the backwards incompatibility of struct attribute types: I don't think we should try to spend all that time trying to avoid that. What we should do instead is publish an actual (pre)release which documents the backwards incompatible change as a helpful hint to users (it doesn't make sense to have a bunch of type aliases everywhere in the long term).

So to review the plan:

1. change the generators to push new common types to `gateway_api::apis::standard::common::*`
2. change the generators to include the new common types in their respective parent types (e.g. `gateway_api::apis::standard::httproute::HTTPRoute`)
3. create a CHANGELOG so we can document the breaking change

If that sounds good to you and we can do it in a way that reduces (not eliminates) changes to the API surface, then that sounds like a plan to me.

dawid-nowak commented:
I think this PR is ready for a proper review. 
* removed processed APIs
* added type generation for experimenal APIs
* added a way to just rename types
* added a mechanism to ignore certain types ... this was needed for experimental UDPRouteSpec and TCPRouteSpec

It would be good to check the mapped types names [here](https://github.com/kubvernor/gateway-api-rs/blob/reducing_common_structs_types/type-reducer/standard_customized_mapped_names.txt) and [here](https://github.com/kubvernor/gateway-api-rs/blob/reducing_common_structs_types/type-reducer/experimental_customized_mapped_names.txt).

shaneutt commented on gateway-api/CHANGELOG.md:
```diff
@@ -0,0 +1,16 @@
+# Changelog
+
+## 0.2.0 
```
Let's call this "Next" and we can move it to a specific version once we're ready to release:

```suggestion
## Next
```

shaneutt commented on gateway-api/CHANGELOG.md:
```diff
@@ -0,0 +1,16 @@
+While with this approach we can significantly reduce the surface of exposed APIs, it is also a breaking change.
+See [issue](https://github.com/kube-rs/gateway-api-rs/issues/38) for more context.
+
+## 0.1.0
```
We already have tags, just not GitHub releases. I've [created a pre-release](https://github.com/kube-rs/gateway-api-rs/releases/tag/v0.16.0) for the latest tag `v0.16.0`, let's use that as the initial release.

```suggestion
## 0.16.0
```

shaneutt commented on gateway-api/src/apis/experimental/common.rs:
```diff
@@ -0,0 +1,388 @@
+/// for the route rule.
+///
+/// Support: Extended
+///
+///
```
Something we can trim in the generator?

dawid-nowak replied:
Yeah, I think I will need to trim all comments out.. They don't make much sense for common types.

shaneutt commented on gateway-api/src/lib.rs:
```diff
@@ -87,29 +88,30 @@ mod tests {
-            supported_kinds: vec![],
-            conditions: vec![Condition {
+        let gw_status = GatewayStatus {
+            addresses: Some(vec![GatewayAddress::default()]),
```
nit: you could use the `..default::Default()` struct update syntax at the bottom instead of initializing this as default.

dawid-nowak replied:
Done

dawid-nowak replied:
Actually, I had to revert this one.. it broke the integration tests.

shaneutt commented on type-reducer/experimental_rename_only_mapped_names.txt:
```diff
@@ -0,0 +1,14 @@
+HTTPRouteRulesMatchesMethod->HTTPMethodMatch
+HTTPRouteRulesBackendRefsFilters->HTTPRouteBackendFilter
+HTTPRouteRequestRedirect->RequestRedirect
+HTTPRouteRulesMatchesPath->PathMatch
```
Is it potentially better to not check in these mapping files because they will be regenerated on every generator run anyway? Or am I misunderstanding?

shaneutt commented on type-reducer/reduce_types.sh:
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+rm -rf ../gateway-api/src/apis/processed
+mkdir ../gateway-api/src/apis/processed
+#cargo run -- --apis-dir ../gateway-api/src/apis/standard --out-dir ../gateway-api/src/apis/processed
```
```suggestion
```

dawid-nowak replied:
I will remove this file, it is no longer needed.

shaneutt commented on type-reducer/standard_rename_only_mapped_names.txt:
```diff
@@ -0,0 +1,14 @@
+HTTPRouteRulesMatchesMethod->HTTPMethodMatch
+HTTPRouteRulesBackendRefsFilters->HTTPRouteBackendFilter
+HTTPRouteRequestRedirect->RequestRedirect
+HTTPRouteRulesMatchesPath->PathMatch
```
Again, wondering if we can avoid checking in the mapping files? WDYT? :thinking:

dawid-nowak replied:
We need this information unfortunately, it shows how to map Kopium dervied type names to new, friendlier names.

shaneutt reviewed (changes requested):
Did a first pass, mostly looking at what is emitted first since it's such a huge diff. I will take another pass to specifically review the `type-reducer` and changes to `update.sh`. :+1: 

Mostly just a few minor things, and some questions.

Also @aharbis @the-wondersmith @kflynn @blinsay @EItanya you're all folks that I think may be using `gateway-api-rs` in some capacity, so you might find this interesting.

dawid-nowak commented:
@shaneutt I have applied the suggestions... I think you might need to re-approve to get it merged.
